### PR TITLE
Concave activity jp

### PR DESCRIPTION
**Pull request by John-Polo**
I wrote an activity/code that uses r.param.scale to calculate vertical concavity on elev_lid792_1m.

### CODE REVIEW

**Comment by John-Polo** [comment]
I read the errors from the P.R. checks and it looks like I had a "," in the JSON that didn't need to be there. Same for an "env" in the .py. Removed them both.

**Comment by wenzeslaus** [comment]
> I had a "," in the JSON that didn't need to be there.

Yes, no trailing commas in JSON (while, confusingly, they need to be present in some cases in Python code formatted with Black).

> Same for an "env" in the .py. Removed them both.

The env is needed because that's how Tangible Landscape manages custom computational region settings within the GRASS session. You need to add it as a parameter to your function.

**Comment by John-Polo** [comment]
It complained about env and I took env out. Now it complains that env is missing.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
This looks good, but you still need to revert removal of the files created by Anna - we want them to stay in the code. I think the following should give you the files back:

```
git checkout upstream/main -- activities/activity_anna_petrasova.py
git checkout upstream/main -- activities/config_anna_petrasova.json
```

The `upstream` part of `upstream/main` requires that you already set up your repository to receive updates from the main (aka upstream) repository (and possibly re-created your fork since you used main branch before there - unrelated to issue at hand but otherwise relevant). You can skip all that and instead of `upstream/main` use `45855021947a32b041ae65cbc0e012a647f694a7` which is the hash of a revision which still contains Anna's files (and it happens to be also the one where she added them, but that's just a coincidence).

File: activities/activity_johnpolo.py
Comment (wenzeslaus) [inline activities/activity_johnpolo.py]: It is perhaps unexpected - perhaps we oversimplified the template - but no custom parameters are allowed for these functions. They are called from Tangible Landscape which does not know what to put there, so these values need to be hardcoded in the function.

To avoid hardcoding at some level, you can create another function (_not_ starting with `run_`) without hardcoded values and call it from `run_topo_par` where you have these hardcoded values. This is little more complicated, so that's why I mentioned the (over)simplification in the template.

File: activities/config_johnpolo.json
Comment (wenzeslaus) [inline activities/config_johnpolo.json]: You probably want `feat_verticalcurv` here. Since it will be hardcoded in the `run_topo_par` function, it will exist when Tangible Landscape runs this function.

Then you will see this in the artifact from Render activities check.

**Review by wenzeslaus — APPROVED** [review]

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
Ha. I meant to request changes, not approve. Sorry for the confusing. But it is in a good shape.

**Comment by wenzeslaus** [comment]
About the formatting complains from the Super-Linter (`Found errors in [black] linter!`), you don't need to fix them manually. Install Black and run it on the file. (The Black does not need to be integrated with GRASS GIS in any way. Black in any Python editor, in Anacodna, or [online](https://black.now.sh/?version=stable) will work.)

**Comment by John-Polo** [comment]
> 
> 
> About the formatting complains from the Super-Linter (`Found errors in [black] linter!`), you don't need to fix them manually. Install Black and run it on the file. (The Black does not need to be integrated with GRASS GIS in any way. Black in any Python editor, in Anacodna, or [online](https://black.now.sh/?version=stable) will work.)

Hi Vaclav,
I used the Black online that you linked to and I copied everything but the trailing blank line over to activities_johnpolo.py, so I should pass Super-Linter, but it still fails. When I look in the Super-Linter report it says that all of my files passed. I don't know what it's complaining about.

File: activities/activity_anna_petrasova.py
Comment (wenzeslaus) [inline activities/activity_anna_petrasova.py]: Can you add back the empty lines here. Then this should disappear from the PR.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]

**Comment by wenzeslaus** [comment]
> ...should pass Super-Linter, but it still fails. When I look in the Super-Linter report it says that all of my files passed.

The output is long, not ideal, so you need to scroll to find things. It is e.g. this line:

```
               return_code, 0, msg=("Running {filename} failed".format(**locals())),
```

But the issue looks like something which happened to me before when I used subsequently different versions of Black, but that's actually not that important. Important is that trailing comma in a line like the one above is taken by current release of Black as an instruction to spread the parameters over multiple lines. It is the default settings and I'm little confused about how exactly it works, but deleting that last (trailing) comma and running Black again should give you code which passes the check.

File: activities/config_johnpolo.json
Comment (John-Polo) [reply activities/config_johnpolo.json]: I substituted feat_verticalcurv for elev...

File: activities/activity_anna_petrasova.py
Comment (John-Polo) [reply activities/activity_anna_petrasova.py]: I added the lines back in.

**Comment by John-Polo** [comment]
This is from the Super-Linter report: 

2021-02-24 14:42:05 [INFO]   File:[/github/workspace/activities/activity_luke_allen.py]
2021-02-24 14:42:05 [ERROR]   Found errors in [black] linter!
2021-02-24 14:42:05 [ERROR]   Error code: 1. Command output:
------
would reformat /github/workspace/activities/activity_luke_allen.py
Oh no! 💥 💔 💥

2021-02-24 14:42:05 [INFO]   File:[/github/workspace/tests/functions.py]
2021-02-24 14:42:06 [ERROR]   Found errors in [black] linter!
2021-02-24 14:42:06 [ERROR]   Error code: 1. Command output:
------
would reformat /github/workspace/tests/functions.py
Oh no! 💥 💔 💥

2021-02-24 14:42:07 [INFO]   File:[/github/workspace/website/render_activities.py]
2021-02-24 14:42:07 [ERROR]   Found errors in [black] linter!
2021-02-24 14:42:07 [ERROR]   Error code: 1. Command output:
------
would reformat /github/workspace/website/render_activities.py
Oh no! 💥 💔 💥

I have no way to fix the last 2, because I don't even know what they are and I'm not going to touch Luke's work. The report says activity_johnpolo.py passed, as well as config_johnpolo.json. This Super-Linter is annoying. :)

File: activities/activity_johnpolo.py
Comment (wenzeslaus) [inline activities/activity_johnpolo.py]: You can delete these instructions and replace them which once sentence description of what you are doing in the script.

See the second point before you delete the text - it relates to my other comment.

File: activities/activity_johnpolo.py
Comment (wenzeslaus) [inline activities/activity_johnpolo.py]: The `overwrite=True,` is actually not needed here because you set that globally with `env["GRASS_OVERWRITE"] = "1"`.

File: activities/activity_johnpolo.py
Comment (wenzeslaus) [inline activities/activity_johnpolo.py]: The parameters of this function need to be the same as in the template. See e.g. Anna's script.

File: activities/activity_johnpolo.py
Comment (wenzeslaus) [inline activities/activity_johnpolo.py]: `select_meth` and `morphmap` need to be hardcoded in the `run_topo_par` function. In general, this would be a bad design, but here it is actually required by the Tangible Landscape API (which wouldn't know what to assign to these parameters).

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
Just the issue with parameters of the `run_...` function remains from the main issues. See the comments.

For some reason restoring Anna's file did not restore the executable permissions which were set on the file. Although they are not needed in Tangible Landscape or in the continuous integrtaion, but let's assume Anna was making use of them and let's try to restore them. The command to do that is (if that doesn't work, let's just ignore this):

```
git update-index --chmod=+x activities/activity_anna_petrasova.py 
```

**Comment by wenzeslaus** [comment]
> I have no way to fix the last 2, because I don't even know what they are and I'm not going to touch Luke's work.

Good call. You are not touching those with you commit, so there is no reason to try to fix them. The problem must be (is) elsewhere.

> The report says activity_johnpolo.py passed, as well as config_johnpolo.json.

As long as these two pass, we can merge this PR.

> This Super-Linter is annoying. :)

I'm actually little frustrated, because I can't figure out where is the issue, specifically why it surfaced only now. But that's fine because it is _not_ related to this PR and I made some changes in #17 to hopefully mitigate the issue.

As far as the specific linter goes - the Python code formatter Black - the idea is that you actually don't need to care about formatting much and you get consistent formatting in the whole project. There are issues and annoyances especially when enforcing it automatically (and Black is actually still in beta so there are even breaking changes here and there), but from my experience, the advantages outweigh that. For one script maybe not, for a project with multiple contributors definitively yes because you don't have to deal with different formatting styles etc. The assumption here is also that eventually, more time is spend reading the code than writing it, so spending some extra time with writing (e.g., in order to minimize distractions when reading it) should pay off in the long run. I hope this helps a little in dealing with these side issues.

File: activities/config_johnpolo.json
Comment (wenzeslaus) [reply activities/config_johnpolo.json]: Great. I checked the result in the artifact and it looks good!

File: activities/activity_johnpolo.py
Comment (John-Polo) [reply activities/activity_johnpolo.py]: I deleted the three statements and wrote a one sentence description. 

In response to the missing execution on the anna_petrasova.py file, I did the following in Git Bash:
john@DESKTOP-7OSMQHM MINGW64 /d/projects_and_data/data/landscape_git/gis714-2021-tangible-landscape (concave_activity_jp)
$ git update-index --chmod=+x activities/activity_anna_petrasova.py

It returned no errors, so I think it worked.

File: activities/activity_johnpolo.py
Comment (John-Polo) [reply activities/activity_johnpolo.py]: I deleted `"overwrite=True"`.

File: activities/activity_johnpolo.py
Comment (John-Polo) [reply activities/activity_johnpolo.py]: I deleted `morphmap` and `select_meth` from `run_topo_par`.

File: activities/activity_johnpolo.py
Comment (John-Polo) [reply activities/activity_johnpolo.py]: I edited the `gs.run.command` by adding: `output="feat_verticalcurv",` and `method="profc",`. This is how I understood what you mean by hardcoding the function. Hopefully, I got it right this time. :)

Comment (wenzeslaus) [reply activities/activity_johnpolo.py]: Yes. That's it!

**Comment by John-Polo** [comment]
Yaayy!!! :D

**Comment by wenzeslaus** [comment]
> In response to the missing execution on the anna_petrasova.py file, ... It returned no errors, so I think it worked.

I don't see it would change in any commits. I will try myself just before merging it with the same command to see how it behaves.

**Review by wenzeslaus — APPROVED** [review]
This is ready to be merged. Congrats!

**Comment by wenzeslaus** [comment]
Great. Close #10 as I assume this fully replaces it. I will merge shortly.

**Comment by John-Polo** [comment]
I closed #10. 
I think. ;)
Am I supposed to close this one (#16) too? There is an option here that says "Close with comment".

**Comment by wenzeslaus** [comment]
> > In response to the missing execution on the anna_petrasova.py file, ... It returned no errors, so I think it worked.
> 
> I don't see it would change in any commits. I will try myself just before merging it with the same command to see how it behaves.

OK, what worked for me was:

```
git update-index --chmod=+x activities/activity_anna_petrasova.py
git commit -m "Restore the original executable permissions"
git restore activities/activity_anna_petrasova.py
```

Notably, the `git commit` is without filename, `-a`, nor it is preceded by `git add`. Note to myself: The `git update-index` change happens in the staging area (it says `update-index`, so I should have expected that). On Linux, there was no actual change of permissions of the filename, so `git add` would revert the permissions back to the original state. The `git restore` then changes the actual permissions.

The file seems to be finally untouched by this commit in the *Files changed*. All the changes to the file will disappear when we do   _**Squash** and merge_.

**Comment by wenzeslaus** [comment]
> I closed #10.

Thanks!

> Am I supposed to close this one (#16) too? There is an option here that says "Close with comment".

Not saying that you want to further work on it and having review with an approval usually indicates that it is ready to be merged (immediately or perhaps after final round of checks finishes). Merging needs to be done by project maintainers (with write access). Merging than closes the PR (closed as _merged_ as opposed to _closed_ as not to be further worked on - perhaps not the best terminology :-).

**Pull request merged**
